### PR TITLE
rxui: TwoWay DataContext and ViewModel Synchronization

### PR DESCRIPTION
**Pull request by worldbeater**
#### What does the pull request do?

This PR synchronizes the `DataContext` property and the `ViewModel` property in `TwoWay` mode. When the `DataContext` property changes, we update the `ViewModel` property. When the `ViewModel` property changes, we update the `DataContext` property. Stumbled upon the lack of `ViewModel->DataContext` synchronization in https://github.com/reactiveui/Sextant/pull/253 Also heard complaints from folks regarding the lack of this synchronization.

#### What is the current behavior?

Currently, when `DataContext` changes, we update the `ViewModel` property, _but not vice versa_.

#### What is the updated/expected behavior with this PR?

Now, when `DataContext` changes, we update the `ViewModel` property, _and vice versa_.

#### How was the solution implemented (if it's not obvious)?

We are using `WhenAnyValue` that uses `.DistinctUntilChanged()` internally.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?

### CODE REVIEW

File: src/Avalonia.ReactiveUI/ReactiveUserControl.cs [hunk 1]
```diff
@@ -20,7 +25,13 @@ public class ReactiveUserControl<TViewModel> : UserControl, IViewFor<TViewModel>
         /// </summary>
         public ReactiveUserControl()
         {
-            DataContextChanged += (sender, args) => ViewModel = DataContext as TViewModel;
+            this.WhenActivated(disposables => { });
+            this.WhenAnyValue(x => x.ViewModel)
+                .Skip(1)
+                .Subscribe(model => DataContext = model);
+            this.WhenAnyValue(x => x.DataContext)
```
Comment (pr8x) [inline src/Avalonia.ReactiveUI/ReactiveUserControl.cs:32]: You can use `ObeservableForProperty` here instead. That way you don't need to `Skip`

**Review by pr8x — COMMENTED** [review]
I also wonder why somebody would set the `ViewModel` property. We could make it readonly?

**Comment by worldbeater** [comment]
> I also wonder why somebody would set the ViewModel property. We could make it readonly?

This is what's happening in https://github.com/reactiveui/Sextant/pull/253#discussion_r509706907 Setting the typed `ViewModel` property and ignoring the untyped `DataContext` property is quite a common practice in ReactiveUI-based applications. E.g. in `ReactiveUI.XamForms`, we are listening to `DataContext` changes and `ViewModel` changes, and update both properties with a new value in order to keep them in sync https://github.com/reactiveui/ReactiveUI/blob/main/src/ReactiveUI.XamForms/ReactiveContentView.cs#L49-L58 Worth noting that the `IViewFor` interface requires the `ViewModel` property to have both a getter and a setter.

Also, there might be cases when an object that implements `IViewFor` doesn't implement `IDataContextProvider`. Not particularly applicable to AvaloniaUI, but anyway we are doing a check in our `RoutedViewHost` control https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.ReactiveUI/RoutedViewHost.cs#L126

File: src/Avalonia.ReactiveUI/ReactiveUserControl.cs (hunk 1 above)
Comment (worldbeater) [reply src/Avalonia.ReactiveUI/ReactiveUserControl.cs:32]: Good catch, thanks! Going to correct this soon.

**Review by danwalmsley — APPROVED** [review]
LGTM

**Pull request merged**